zipline/finance/slippage.py

```diff
@@ -252,3 +349,124 @@ def process_order(self, data, order):
+    """
+    pass
+
+
+class FixedFutureSlippage(FixedSlippageBase, FutureSlippageModel):
+    """
+    Model slippage for futures as a fixed spread.
+
+    Parameters
+    ----------
+    spread : float, optional
+        spread / 2 will be added to buys and subtracted from sells.
+    """
+    pass
+
+
+class WithWindowData(object):
+    """
+    Mixin class for slippage models requiring average daily volume and
+    annualized daily volatility.
+    """
+
+    def __init__(self):
+        super(WithWindowData, self).__init__()
+        self._window_data_cache = ExpiringCache()
+
+    def _get_window_data(self, data, asset, window_length):
+        """
+        Internal utility method to return the trailing 20-day mean volume and
+        volatility of close prices for a specific asset.
+
+        Parameters
+        ----------
+        data : The BarData from which to fetch the daily windows.
+        asset : The Asset whose data we are fetching.
+
+        Returns
+        -------
+        (mean volume, volatility)
+        """
+        try:
+            values = self._window_data_cache.get(asset, data.current_session)
+        except KeyError:
+            # Add a day because we wan 'window_length' complete days.
+            volume_history = data.history(
+                asset, 'volume', window_length + 1, '1d',
+            )
+            close_history = data.history(
+                asset, 'close', window_length + 1, '1d',
+            )
+            values = {
+                'volume': volume_history[:-1].mean(),
+                'close': close_history[:-1].pct_change().std() * SQRT_252,
+            }
+            self._window_data_cache.set(asset, values, data.current_session)
+
+        return values['volume'], values['close']
+
+
+class VolatilityVolumeShare(WithWindowData, FutureSlippageModel):
+
+    def __init__(self, volume_limit):
+        super(VolatilityVolumeShare, self).__init__()
+        self.volume_limit = volume_limit
+
+    def process_order(self, data, order):
+        if order.open_amount == 0:
+            return None, None
+
+        minute_data = data.current(order.asset, ['volume', 'high', 'low'])
+        volume = minute_data['volume']
+        if not volume:
+            return None, None
+
+        # Price to use is the average of the minute bar's open and close.
+        price = np.mean([minute_data['high'], minute_data['low']])
+
+        eta = ROOT_SYMBOL_TO_ETA[order.asset.root_symbol]
+        mean_volume, volatility = self._get_window_data(data, order.asset, 20)
+
+        if mean_volume == 0:
+            # If this is the first day the contract exists and there is no
+            # volume history, default to an impact of 10 bps.
+            simulated_impact = price * (10.0 / 10000)
+
+            impacted_price = \
+                price + math.copysign(simulated_impact, order.direction)
+        else:
+            txn_volume = int(
```

Comment:
I'm wondering if we should refactor this method so that it can be shared between VolatilityVolumeShare and MarketImpact and this else block can be a call to get simulated_impact because that appears to be the only place this logic differs.